### PR TITLE
Sorted out efficiency and correctly labelling of quarter abbreviations in the use date dialog

### LINKED ISSUE #7027: Quarter Abbr produced from Use Date dialog should be a factor not a character

### PR DESCRIPTION
**Pull request by shadrackkibet**
This PR branches from #7030 hence fixes #7027.

@N-thony here are the R code changes needed for this to work properly.  I have sorted out the following issues.

-  Moved the function `get_quarter_label()`  into the list of stand-alone functions. This is because this function doesn't depend on objects in the `DataSheet`.
- Ensured that levels are in the right order by explicitly specifying levels `factor(x = qtr, levels = unique(qtr))`.
- Improved efficiency by avoiding the `for` loop.
- Removed repetition.

@dannyparsons please check when you have time.

@africanmathsinitiative/developers this is ready for review.

### CODE REVIEW

**Comment by N-thony** [comment]
> This PR branches from #7030 hence fixes #7027.
> 
> @N-thony here are the R code changes needed for this to work properly. I have sorted out the following issues.
> 
> * Moved the function `get_quarter_label()`  into the list of stand-alone functions. This is because this function doesn't depend on objects in the `DataSheet`.
> * Ensured that levels are in the right order by explicitly specifying levels `factor(x = qtr, levels = unique(qtr))`.
> * Improved efficiency by avoiding the `for` loop.
> * Removed repetition.
> 
> @dannyparsons please check when you have time.
> 
> @africanmathsinitiative/developers this is ready for review.

@shadrackkibet thanks. I also checked on my side, it works fine. :)

File: instat/static/InstatObject/R/stand_alone_functions.R [hunk 1]
```diff
@@ -2507,4 +2507,17 @@ slopegraph <- function(data, x, y, colour, data_label = NULL,
...
-}
\ No newline at end of file
+}
+
+# Returns a three-letter string representing a specific quarter in a year (e.g. "JFM", "AMJ" etc.). 
+get_quarter_label <-   function(quarter, start_month){
+  if (!start_month %in% 1:12) stop(start_month, " is an invalid start month")
+  if (!all(quarter %in% 1:4)) stop("All values must be in range of 1:4")
+  if (start_month == 1) {
+    mabb <- substr(month.abb, 1, 1)
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/stand_alone_functions.R:2517]: Will `substr(month.abb, 1, 1)` just return "J"?
Or did I misunderstand?
Thanks

Comment (lloyddewit) [reply instat/static/InstatObject/R/stand_alone_functions.R:2517]: OK, got it, now I understand
thanks

Comment (lloyddewit) [inline instat/static/InstatObject/R/stand_alone_functions.R:2517]: The `mabb` calculated on line 2519 also works when `start_month` is `1`.
Should we remove this `if` statement?

File: instat/static/InstatObject/R/stand_alone_functions.R [hunk 2]
```diff
@@ -2507,4 +2507,17 @@ slopegraph <- function(data, x, y, colour, data_label = NULL,
...
+  } else {
+    mabb <- rep(substr(month.abb, 1, 1), times = 2)[start_month:(12 + start_month)]
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/stand_alone_functions.R:2519]:
```suggestion
    mabb <- rep(substr(month.abb, 1, 1), times = 2)[start_month:(11 + start_month)]
```

File: instat/static/InstatObject/R/stand_alone_functions.R [hunk 3]
```diff
@@ -2507,4 +2507,17 @@ slopegraph <- function(data, x, y, colour, data_label = NULL,
...
+  }
+  qtr <- ifelse(quarter == 1, paste(mabb[1:3], collapse = ""), ifelse(quarter == 2, paste(mabb[4:6], collapse = ""), ifelse(quarter == 3, paste(mabb[7:9], collapse = ""), paste(mabb[10:12], collapse = ""))))
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/stand_alone_functions.R:2521]: If we calculate a `start_pos` (e.g. `start_pos <= 1 + ((quarter-1) * 3)`), then we could shorten this line, e.g
```
qtr <- paste(mabb[start_pos:start_pos+2], collapse = "")
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]

File: instat/static/InstatObject/R/stand_alone_functions.R [hunk 4]
```diff
@@ -2507,4 +2507,17 @@ slopegraph <- function(data, x, y, colour, data_label = NULL,
...
     ggplot2::geom_label(ggplot2::aes_string(label = Ndata_label), size = data_text_size, label.padding = unit(data_label_padding, "lines"),
                label.size = data_label_line_size, colour = data_text_colour, fill = data_label_fill_colour)
-}
\ No newline at end of file
+}
+
+# Returns a three-letter string representing a specific quarter in a year (e.g. "JFM", "AMJ" etc.). 
+get_quarter_label <-   function(quarter, start_month){
+  if (!start_month %in% 1:12) stop(start_month, " is an invalid start month")
+  if (!all(quarter %in% 1:4)) stop("All values must be in range of 1:4")
```
Comment (lloyddewit) [inline instat/static/InstatObject/R/stand_alone_functions.R:2515]: Make the messages more consistent and useful?
```suggestion
  if (!start_month %in% 1:12) stop(start_month, " is an invalid start month, must be in range of 1:12")
  if (!all(quarter %in% 1:4)) stop(quarter, " is an invalid quarter, must be in range of 1:4")
```

File: instat/static/InstatObject/R/stand_alone_functions.R (hunk 1 above)
Comment (shadrackkibet) [reply instat/static/InstatObject/R/stand_alone_functions.R:2517]: Absolutely! thanks for the insights on simplification.

File: instat/static/InstatObject/R/stand_alone_functions.R (hunk 3 above)
Comment (shadrackkibet) [reply instat/static/InstatObject/R/stand_alone_functions.R:2521]: Yes, I got the idea. I will simplify it further. Thanks for pointing out this.

**Comment by shadrackkibet** [comment]
@lloyddewit thanks for the review. Your insights were useful in simplifying the code.

**Review by lloyddewit — APPROVED** [review]
@shadrackkibet Thank you! looks good to me

**Comment by lloyddewit** [comment]
@shadrackkibet who's the best person for final approval?

**Comment by rdstern** [comment]
I am happy to approve the changes to the quarter function.  I am a bit concerned about the look of other controls in the dialogue?  Is this something that will be ok in the merged version?
image

**Review by rdstern — APPROVED** [review]
@shadrackkibet I am approving, given @dannyparsons is happy.  I am still concerned at the lack of visible checkboxes in the dialogue and hope that won't be the case in the merged version.

**Comment by shadrackkibet** [comment]
Great, I am going to merge the `quarter` changes now.

@africanmathsinitiative/developers could someone help with sorting out the visibility of some checkboxes in this dialog? This is something in the current master.

**Pull request merged**